src/axom/primal/tests/primal_nurbs_patch.cpp

```diff
@@ -0,0 +1,1322 @@
+    PointType {8, 0, 0}, PointType {8, 4,  0}, PointType {8, 8,  0}, PointType {8, 12, 0}};
+
+  double weights[5 * 4] = {
+    1.0, 2.0, 3.0, 2.0,
+    2.0, 3.0, 4.0, 3.0,
+    3.0, 4.0, 5.0, 4.0,
+    4.0, 5.0, 6.0, 5.0,
+    5.0, 6.0, 7.0, 6.0};
+  // clang-format on
+
+  NURBSPatchType nPatch(controlPoints, weights, npts_u, npts_v, degree_u, degree_v);
+
+  // isocurve_u should *fix* a value of u, returning a curve parameterized by v
+  EXPECT_EQ(nPatch.isocurve_u(0.5).getDegree(), degree_v);
+  EXPECT_EQ(nPatch.isocurve(0.5, 0).getDegree(), degree_v);
+
+  // isocurve_v should *fix* a value of v, returning a curve parameterized by u
+  EXPECT_EQ(nPatch.isocurve_v(0.5).getDegree(), degree_u);
+  EXPECT_EQ(nPatch.isocurve(0.5, 1).getDegree(), degree_u);
+
+  // Loop over the parameter space of the surface,
+  //  and check that `evalaute` matches the results of the two isocurve methods
+  for(double u = 0; u <= 1; u += 0.1)
+  {
+    for(double v = 0; v <= 1; v += 0.1)
+    {
+      auto pt = nPatch.evaluate(u, v);
+      auto pt_u = nPatch.isocurve_u(u).evaluate(v);
+      auto pt_v = nPatch.isocurve_v(v).evaluate(u);
+
+      for(int i = 0; i < DIM; ++i)
+      {
+        EXPECT_NEAR(pt[i], pt_u[i], 1e-10);
+        EXPECT_NEAR(pt[i], pt_v[i], 1e-10);
+        EXPECT_NEAR(pt_u[i], pt_v[i], 1e-10);
+      }
+    }
+  }
+}
+
+//------------------------------------------------------------------------------
+TEST(primal_nurbspatch, first_second_derivatives)
+{
+  SLIC_INFO("Testing NURBS Patch derivative evaluation");
+
+  const int DIM = 3;
+  using CoordType = double;
+  using PointType = primal::Point<CoordType, DIM>;
+  using VectorType = primal::Vector<CoordType, DIM>;
+  using NURBSPatchType = primal::NURBSPatch<CoordType, DIM>;
+
+  const int npts_u = 5;
+  const int npts_v = 4;
+
+  const int degree_u = 3;
+  const int degree_v = 2;
+
+  // clang-format off
+  PointType controlPoints[5 * 4] = {
+    PointType {0, 0, 0}, PointType {0, 4,  0}, PointType {0, 8, -3}, PointType {0, 12, 0},
+    PointType {2, 0, 6}, PointType {2, 4,  0}, PointType {2, 8,  0}, PointType {2, 12, 0},
+    PointType {4, 0, 0}, PointType {4, 4,  0}, PointType {4, 8,  3}, PointType {4, 12, 0},
+    PointType {6, 0, 0}, PointType {6, 4, -3}, PointType {6, 8,  0}, PointType {6, 12, 0},
+    PointType {8, 0, 0}, PointType {8, 4,  0}, PointType {8, 8,  0}, PointType {8, 12, 0}};
+
+  double weights[5 * 4] = {
+    1.0, 2.0, 3.0, 2.0,
+    2.0, 3.0, 4.0, 3.0,
+    3.0, 4.0, 5.0, 4.0,
+    4.0, 5.0, 6.0, 5.0,
+    5.0, 6.0, 7.0, 6.0};
+  // clang-format on
+
+  NURBSPatchType nPatch(controlPoints, weights, npts_u, npts_v, degree_u, degree_v);
+
+  // Loop over the parameter space of the surface,
+  //  and check that `evalauteDerivatives` matches the results of the two isocurve methods
+  for(double u = 0; u <= 1; u += 0.1)
+  {
+    for(double v = 0; v <= 1; v += 0.1)
+    {
+      axom::Array<VectorType, 2> ders;
```

Comment:
This `VectorType` refers to a geometric `primal::Vector`, not a `std::vector` array.

Reply:
There is a `VectorType` defined within the NURBSPatch class, do you think it should be defined like this throughout the test? Or maybe instead of the `using` format, just define use it as `NURBSPatchType::VectorType` throughout.

```C++
const int DIM = 3;
using CoordType = double;
using NURBSPatchType = primal::NURBSPatch<CoordType, DIM>;
using VectorType = NURBSPatchType::VectorType;
```

Reply:
I was thinking it might make sense to have something like 
```cpp
using Derivatives = axom::Array<VectorType>;  // or DerivativesType (?)
```
in `NURBSCurve` and/or `NURBSPatch`

It's not strictly necessary though.

Reply:
Ah, I see what you mean, I actually like the idea of having a specific `DerivativeType` in those classes, but I would want that to be reflected in all the other curved objects. Unless it's worth it to change all those classes now, I might save it for a later PR that also addresses some other things that I think would be good to rework among all of them (like array accessing).

Reply:
Thanks -- that works for me